politeiawww/dcc.go

```diff
@@ -970,3 +1022,251 @@ func dccStatusInSlice(arr []cms.DCCStatusT, status cms.DCCStatusT) bool {
+		}
+	}
+	vdr, err := p.cmsVoteDetails(vd.Token)
+	if err != nil {
+		return nil, err
+	}
+	dcc, err := p.getDCC(vd.Token)
+	if err != nil {
+		if err == cache.ErrRecordNotFound {
+			err = www.UserError{
+				ErrorCode: cms.ErrorStatusDCCNotFound,
+			}
+		}
+		return nil, err
+	}
+
+	// Only allow voting on All Vote DCC proposals
+	if dcc.Status != cms.DCCStatusAllVote {
+		return nil, www.UserError{
+			ErrorCode: cms.ErrorStatusInvalidDCCVoteStatus,
+		}
+	}
+
+	bb, err := p.getBestBlock()
+	if err != nil {
+		return nil, err
+	}
+
+	// Check to make sure that the Vote hasn't ended yet.
+	if uint64(vdr.StartVoteReply.EndHeight) < bb {
+		return nil, www.UserError{
+			ErrorCode: cms.ErrorStatusDCCVoteEnded,
+		}
+	}
+
+	challenge, err := util.Random(pd.ChallengeSize)
+	if err != nil {
+		return nil, err
+	}
+
+	vote := convertCastVoteFromCMS(vd)
+	for _, voteOption := range vdr.StartVote.Vote.Options {
+		if voteOption.Id == vd.Vote {
+			vote.VoteBit = strconv.FormatUint(voteOption.Bits, 16)
+			break
+		}
+	}
+	payload, err := cmsplugin.EncodeCastVote(vote)
+	if err != nil {
+		return nil, err
+	}
+	pc := pd.PluginCommand{
+		Challenge: hex.EncodeToString(challenge),
+		ID:        cmsplugin.ID,
+		Command:   cmsplugin.CmdCastVote,
+		CommandID: cmsplugin.CmdCastVote,
+		Payload:   string(payload),
+	}
+
+	responseBody, err := p.makeRequest(http.MethodPost,
+		pd.PluginCommandRoute, pc)
+	if err != nil {
+		return nil, err
+	}
+
+	var reply pd.PluginCommandReply
+	err = json.Unmarshal(responseBody, &reply)
+	if err != nil {
+		return nil, fmt.Errorf("Could not unmarshal "+
+			"PluginCommandReply: %v", err)
+	}
+
+	// Verify the challenge.
+	err = util.VerifyChallenge(p.cfg.Identity, challenge, reply.Response)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode plugin reply
+	_, err = cmsplugin.DecodeCastVoteReply([]byte(reply.Payload))
+	if err != nil {
+		return nil, err
+	}
+
+	return &cms.VoteDCCReply{}, nil
+}
+
+func (p *politeiawww) allVoteDCC(token, signature string, u *user.User) error {
+	log.Tracef("allVoteDCC: %v %v", token, u.PublicKey())
+
+	userWeights, err := p.getCMSUserWeights()
+	if err != nil {
+		return err
+	}
+
+	cmsUserWeights := make([]cmsplugin.UserWeight, 0, len(userWeights))
+
+	for id, weight := range userWeights {
+		cmsuw := cmsplugin.UserWeight{
+			UserID: id,
+			Weight: weight,
+		}
+		cmsUserWeights = append(cmsUserWeights, cmsuw)
+	}
+
+	duration := 2016
+	quorum := 10
+	pass := 60
+	sv := cmsplugin.StartVote{
+		UserWeights: cmsUserWeights,
+		PublicKey:   u.PublicKey(),
+		Signature:   signature,
+		Vote: cmsplugin.Vote{
+			Token:            token,
+			Mask:             0x03, // bit 0 no, bit 1 yes
+			Duration:         uint32(duration),
+			QuorumPercentage: uint32(quorum),
+			PassPercentage:   uint32(pass),
+			Options: []cmsplugin.VoteOption{
+				{
+					Id:          cmsplugin.DCCDisapprovalString,
+					Description: "Don't approve DCC",
+					Bits:        0x01,
```

Comment:
You're defining the vote option using vote bits here but the the user is not actually casting a vote using the vote bit, they're casting the vote using the vote id.  If you define a vote option with a vote bit then the user needs to use it to cast their vote.  

Vote bits are used in pi so that proposal votes can be expanded to multiple choice without any issues, where as comment votes use a simple "upvote" or "downvote" to cast a vote.